substrate-node/pallets/pallet-smart-contract/src/billing.rs

```diff
@@ -0,0 +1,747 @@
+use crate::*;
+use frame_support::{
+    dispatch::{DispatchErrorWithPostInfo, DispatchResultWithPostInfo},
+    ensure,
+    traits::{Currency, ExistenceRequirement, LockableCurrency, OnUnbalanced, WithdrawReasons},
+};
+use frame_system::offchain::{SendSignedTransaction, SignMessage, Signer};
+use sp_core::Get;
+use sp_runtime::{
+    traits::{CheckedAdd, CheckedSub, Convert, Zero},
+    DispatchResult, Perbill, SaturatedConversion,
+};
+use sp_std::vec::Vec;
+
+impl<T: Config> Pallet<T> {
+    pub fn bill_conttracts_for_block(block_number: T::BlockNumber) {
+        // Let offchain worker check if there are contracts on
+        // billing loop at current index and try to bill them
+        let index = Self::get_billing_loop_index_from_block_number(block_number);
+
+        let contract_ids = ContractsToBillAt::<T>::get(index);
+        if contract_ids.is_empty() {
+            log::info!(
+                "No contracts to bill at block {:?}, index: {:?}",
+                block_number,
+                index
+            );
+            return;
+        }
+
+        log::info!(
+            "{:?} contracts to bill at block {:?}",
+            contract_ids,
+            block_number
+        );
+
+        for contract_id in contract_ids {
+            if let Some(c) = Contracts::<T>::get(contract_id) {
+                if let types::ContractData::NodeContract(node_contract) = c.contract_type {
+                    // Is there IP consumption to bill?
+                    let bill_ip = node_contract.public_ips > 0;
+
+                    // Is there CU/SU consumption to bill?
+                    // No need for preliminary call to contains_key() because default resource value is empty
+                    let bill_cu_su = !NodeContractResources::<T>::get(contract_id).used.is_empty();
+
+                    // Is there NU consumption to bill?
+                    // No need for preliminary call to contains_key() because default amount_unbilled is 0
+                    let bill_nu =
+                        ContractBillingInformationByID::<T>::get(contract_id).amount_unbilled > 0;
+
+                    // Don't bill if no IP/CU/SU/NU to be billed
```

Comment:
In my view the comment helps to understand the conclusion of the flow, to conclude what was done before
and could be useful for identifying logic issue which is not obvious at first sight since we use a sequence of logical operators (even if variable names are explicit)